<a href="https://colab.research.google.com/github/BernardOforiBoateng/English-Premier-League/blob/main/DSCI401_PROJECT.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Exploratory Data Analysis on the English Premier League

In [ ]:
#Importing the required libraries
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns
import plotly.express as px

In [ ]:
#loading the dataset
df = pd.read_csv("https://raw.githubusercontent.com/BernardOforiBoateng/Data_Wrangling_and_EDA_DSCI401/main/Project/premier-league-matches.csv")
df.head()

Season_End_Year  Wk        Date            Home  HomeGoals  AwayGoals  \
0             1993   1  1992-08-15   Coventry City          2          1   
1             1993   1  1992-08-15    Leeds United          2          1   
2             1993   1  1992-08-15   Sheffield Utd          2          1   
3             1993   1  1992-08-15  Crystal Palace          3          3   
4             1993   1  1992-08-15         Arsenal          2          4   

             Away FTR  
0   Middlesbrough   H  
1       Wimbledon   H  
2  Manchester Utd   H  
3       Blackburn   D  
4    Norwich City   A

In [ ]:
df.tail()

Season_End_Year  Wk        Date            Home  HomeGoals  AwayGoals  \
12021             2023  38  2023-05-28         Everton          1          0   
12022             2023  38  2023-05-28  Leicester City          2          1   
12023             2023  38  2023-05-28     Aston Villa          2          1   
12024             2023  38  2023-05-28    Leeds United          1          4   
12025             2023  38  2023-05-28       Brentford          1          0   

                  Away FTR  
12021      Bournemouth   H  
12022         West Ham   H  
12023         Brighton   H  
12024        Tottenham   A  
12025  Manchester City   H

In [ ]:
#Check the shape of the dataframe
df.shape

(12026, 8)

In [ ]:
#Check the info of the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12026 entries, 0 to 12025
Data columns (total 8 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Season_End_Year  12026 non-null  int64 
 1   Wk               12026 non-null  int64 
 2   Date             12026 non-null  object
 3   Home             12026 non-null  object
 4   HomeGoals        12026 non-null  int64 
 5   AwayGoals        12026 non-null  int64 
 6   Away             12026 non-null  object
 7   FTR              12026 non-null  object
dtypes: int64(4), object(4)
memory usage: 751.8+ KB


In [ ]:
#Rename Columns
df.rename(columns={"Wk": "Week", "Home": "HomeTeam", "Away": "AwayTeam", "FTR" :"FullTimeResult"}, inplace=True)
df.head()

Season_End_Year  Week        Date        HomeTeam  HomeGoals  AwayGoals  \
0             1993     1  1992-08-15   Coventry City          2          1   
1             1993     1  1992-08-15    Leeds United          2          1   
2             1993     1  1992-08-15   Sheffield Utd          2          1   
3             1993     1  1992-08-15  Crystal Palace          3          3   
4             1993     1  1992-08-15         Arsenal          2          4   

         AwayTeam FullTimeResult  
0   Middlesbrough              H  
1       Wimbledon              H  
2  Manchester Utd              H  
3       Blackburn              D  
4    Norwich City              A

In [ ]:
#Create a points column
# A win is 3 points, draw is 1 point each
df['Points']=3
df.loc[df['FullTimeResult']=='D', 'Points']=1
df.head()

Season_End_Year  Week        Date        HomeTeam  HomeGoals  AwayGoals  \
0             1993     1  1992-08-15   Coventry City          2          1   
1             1993     1  1992-08-15    Leeds United          2          1   
2             1993     1  1992-08-15   Sheffield Utd          2          1   
3             1993     1  1992-08-15  Crystal Palace          3          3   
4             1993     1  1992-08-15         Arsenal          2          4   

         AwayTeam FullTimeResult  Points  
0   Middlesbrough              H       3  
1       Wimbledon              H       3  
2  Manchester Utd              H       3  
3       Blackburn              D       1  
4    Norwich City              A       3

In [ ]:
#Lets check the total number of games played in each season
df['Season_End_Year'].value_counts()

1993    462
1995    462
1994    462
2010    380
2022    380
2021    380
2020    380
2019    380
2018    380
2017    380
2016    380
2015    380
2014    380
2013    380
2012    380
2011    380
2008    380
2009    380
2007    380
2006    380
2005    380
2004    380
2003    380
2002    380
2001    380
2000    380
1999    380
1998    380
1997    380
1996    380
2023    380
Name: Season_End_Year, dtype: int64

As we can see the premier league originally started with 22 teams for the first three seasons after which it was changed to 20 teams.
The Premier League reduced the number of teams from a 22 team league with 462 matches a season to  a 20 team league to combat fixture congestion and player fatigue, to improve the quality of matches and to increase its competitiveness both domestically and internationally.

In [ ]:
#Get unique team names from the dataset
home_teams = df['HomeTeam'].unique()
away_teams = df['AwayTeam'].unique()
all_teams = pd.Series(list(set(home_teams) | set(away_teams)))

# Display unique team names
all_teams


0          Sunderland
1        Leeds United
2         Bournemouth
3     Oldham Athletic
4     Nott'ham Forest
5        Swindon Town
6       Middlesbrough
7                 QPR
8        Derby County
9       Bradford City
10          Liverpool
11       Huddersfield
12      Sheffield Utd
13          West Brom
14      Newcastle Utd
15         Stoke City
16             Fulham
17          Hull City
18            Chelsea
19     Sheffield Weds
20        Aston Villa
21           West Ham
22       Norwich City
23        Southampton
24     Leicester City
25     Manchester Utd
26       Ipswich Town
27         Portsmouth
28            Burnley
29          Tottenham
30             Bolton
31      Coventry City
32          Blackburn
33            Reading
34           Barnsley
35            Everton
36           Brighton
37            Arsenal
38       Swansea City
39    Birmingham City
40       Charlton Ath
41     Crystal Palace
42       Cardiff City
43          Wimbledon
44            Watford
45     Wig

Now the Premier League much like most European Football leagues are dominated by a select few who have the financial power to buy the best players and employ the best coaches. Lets check the trend of how the "Big Six" teams have perfomed over the years.

In [ ]:
#Rename specific team name in the dataset to full name
df['HomeTeam'] = df['HomeTeam'].replace("Nott'ham Forest", "Nottingham Forest")
df['AwayTeam'] = df['AwayTeam'].replace("Nott'ham Forest", "Nottingham Forest")

#Get unique team names from the dataset
home_teams = df['HomeTeam'].unique()
away_teams = df['AwayTeam'].unique()
all_teams = pd.Series(list(set(home_teams) | set(away_teams)))

#Create a new DataFrame to store total points for each team at the end of each season
total_points_df = pd.DataFrame(index=all_teams, columns=range(1993, 2024))
total_points_df = total_points_df.fillna(0)

#Iterate through each row in the original DataFrame
for index, row in df.iterrows():
    season_end_year = row['Season_End_Year']
    home_team = row['HomeTeam']
    away_team = row['AwayTeam']
    full_time_result = row['FullTimeResult']

    # Update points based on match outcome
    if full_time_result == 'H':  #Home team won
        total_points_df.at[home_team, season_end_year] += 3
    elif full_time_result == 'A':  #Away team won
        total_points_df.at[away_team, season_end_year] += 3
    else:  #Draw
        total_points_df.at[home_team, season_end_year] += 1
        total_points_df.at[away_team, season_end_year] += 1

#Display the total points for each team at the end of each season
total_points_df

1993  1994  1995  1996  1997  1998  1999  2000  2001  2002  \
Sunderland            0     0     0     0    40     0     0    58    57    40   
Leeds United         51    70    73    43    46    59    67    69    68    66   
Bournemouth           0     0     0     0     0     0     0     0     0     0   
Oldham Athletic      49    40     0     0     0     0     0     0     0     0   
Swindon Town          0    30     0     0     0     0     0     0     0     0   
Middlesbrough        44     0     0    43    42     0    51    52    42    45   
QPR                  63    60    60    33     0     0     0     0     0     0   
Derby County          0     0     0     0    46    55    52    38    42    30   
Bradford City         0     0     0     0     0     0     0    36    26     0   
Liverpool            59    60    74    71    68    65    54    67    69    80   
Huddersfield          0     0     0     0     0     0     0     0     0     0   
Sheffield Utd        52    42     0     0     0     0     0     0     0     0   
West Brom             0     0     0     0     0     0     0     0     0     0   
Newcastle Utd         0    77    72    78    68    44    46    52    51    71   
Stoke City            0     0     0     0     0     0     0     0     0     0   
Fulham                0     0     0     0     0     0     0     0     0    44   
Hull City             0     0     0     0     0     0     0     0     0     0   
Chelsea              56    51    54    50    59    63    75    65    61    64   
Sheffield Weds       59    64    51    40    57    44    46    31     0     0   
Aston Villa          74    57    48    63    61    57    55    58    54    50   
West Ham              0    52    50    51    42    56    57    55    42    53   
Norwich City         72    53    43     0     0     0     0     0     0     0   
Southampton          50    43    54    38    41    48    41    44    52    45   
Leicester City        0     0    29     0    47    53    49    55    48    28   
Manchester Utd       84    92    88    82    75    77    79    91    80    77   
Ipswich Town         52    43    27     0     0     0     0     0    66    36   
Portsmouth            0     0     0     0     0     0     0     0     0     0   
Burnley               0     0     0     0     0     0     0     0     0     0   
Tottenham            59    45    62    61    46    44    47    53    49    50   
Bolton                0     0     0    29     0    40     0     0     0    40   
Coventry City        52    56    50    38    41    52    42    44    34     0   
Nottingham Forest    40     0    77    58    34     0    30     0     0     0   
Blackburn            71    84    89    61    42    58    35     0     0    46   
Reading               0     0     0     0     0     0     0     0     0     0   
Barnsley              0     0     0     0     0    35     0     0     0     0   
Everton              53    44    50    61    42    40    43    50    42    43   
Brighton              0     0     0     0     0     0     0     0     0     0   
Arsenal              56    71    51    63    68    78    78    73    70    87   
Swansea City          0     0     0     0     0     0     0     0     0     0   
Birmingham City       0     0     0     0     0     0     0     0     0     0   
Charlton Ath          0     0     0     0     0     0    36     0    52    44   
Crystal Palace       49     0    45     0     0    33     0     0     0     0   
Cardiff City          0     0     0     0     0     0     0     0     0     0   
Wimbledon            54    65    56    41    56    44    42    33     0     0   
Watford               0     0     0     0     0     0     0    24     0     0   
Wigan Athletic        0     0     0     0     0     0     0     0     0     0   
Manchester City      57    45    49    38     0     0     0     0    34     0   
Wolves                0     0     0     0     0     0     0     0     0     0   
Blackpool             0     0     0     0     0   

In [ ]:
#List of the big six teams
big_six_teams = ['Manchester Utd', 'Liverpool', 'Arsenal', 'Chelsea', 'Manchester City', 'Tottenham']

#Get  data for the big six teams
big_six_data = total_points_df.loc[big_six_teams]

#Transpose the DataFrame for better visualization
big_six_data = big_six_data.transpose()

#Create an interactive line plot using Plotly Express
fig = px.line(big_six_data, x=big_six_data.index, y=big_six_teams,
              labels={'index': 'Season', 'value': 'Total Points'},
              title='Performance of Big Six Teams Over the Years',
              line_shape="linear",
              )

#Add gridlines and show the plot
fig.update_layout(showlegend=True, xaxis=dict(showgrid=True), yaxis=dict(showgrid=True))
fig.show()


Teams with the Most Home and Away Points

In [ ]:
#Calculate home and away points for each team in each season
home_points = df[df['FullTimeResult'].isin(['H', 'D'])].groupby(['Season_End_Year', 'HomeTeam'])['Points'].sum().reset_index()
away_points = df[df['FullTimeResult'].isin(['A', 'D'])].groupby(['Season_End_Year', 'AwayTeam'])['Points'].sum().reset_index()

#Team with the most home points for each season
most_home_points = home_points.groupby('Season_End_Year').apply(lambda x: x.loc[x['Points'].idxmax()]).reset_index(drop=True)

#Team with the most away points for each season
most_away_points = away_points.groupby('Season_End_Year').apply(lambda x: x.loc[x['Points'].idxmax()]).reset_index(drop=True)

In [ ]:
#Display the results
print("Team with the Most Home Points Each Season:")
most_home_points

Team with the Most Home Points Each Season:


Season_End_Year         HomeTeam  Points
0              1993   Manchester Utd      47
1              1994   Manchester Utd      48
2              1995        Blackburn      53
3              1996    Newcastle Utd      52
4              1997    Newcastle Utd      42
5              1998          Arsenal      47
6              1999          Arsenal      47
7              2000   Manchester Utd      49
8              2001          Arsenal      48
9              2002        Liverpool      41
10             2003   Manchester Utd      50
11             2004          Arsenal      49
12             2005          Chelsea      47
13             2006          Chelsea      55
14             2007   Manchester Utd      47
15             2008   Manchester Utd      52
16             2009   Manchester Utd      50
17             2010          Chelsea      52
18             2011   Manchester Utd      55
19             2012  Manchester City      55
20             2013   Manchester Utd      48
21             2014  Manchester City      52
22             2015          Chelsea      49
23             2016   Leicester City      42
24             2017        Tottenham      53
25             2018  Manchester City      50
26             2019  Manchester City      54
27             2020        Liverpool      55
28             2021  Manchester City      41
29             2022        Liverpool      49
30             2023  Manchester City      52

In [ ]:
print("Team with the Most Away Points Each Season:")
most_away_points


Team with the Most Away Points Each Season:


Season_End_Year         AwayTeam  Points
0              1993   Manchester Utd      37
1              1994   Manchester Utd      44
2              1995        Blackburn      36
3              1996   Manchester Utd      33
4              1997   Manchester Utd      34
5              1998   Manchester Utd      34
6              1999          Chelsea      33
7              2000   Manchester Utd      42
8              2001   Manchester Utd      33
9              2002          Arsenal      47
10             2003   Manchester Utd      33
11             2004          Arsenal      41
12             2005          Chelsea      48
13             2006   Manchester Utd      39
14             2007   Manchester Utd      42
15             2008          Chelsea      42
16             2009          Chelsea      44
17             2010   Manchester Utd      36
18             2011          Arsenal      31
19             2012   Manchester Utd      42
20             2013   Manchester Utd      41
21             2014          Arsenal      35
22             2015          Chelsea      38
23             2016   Leicester City      39
24             2017          Chelsea      42
25             2018  Manchester City      50
26             2019        Liverpool      44
27             2020        Liverpool      44
28             2021  Manchester City      45
29             2022  Manchester City      46
30             2023          Arsenal      39

Teams with the Least amount of points in each season

In [ ]:
#Calculate home and away points for each team in each season
home_points = df[df['FullTimeResult'].isin(['H', 'D'])].groupby(['Season_End_Year', 'HomeTeam'])['Points'].sum().reset_index()
away_points = df[df['FullTimeResult'].isin(['A', 'D'])].groupby(['Season_End_Year', 'AwayTeam'])['Points'].sum().reset_index()

#Merge home and away points, filling NaN with 0
total_points = pd.merge(home_points, away_points, how='outer', left_on=['Season_End_Year', 'HomeTeam'], right_on=['Season_End_Year', 'AwayTeam'])
total_points = total_points.fillna(0)

#Calculate total points for each team in each season
total_points['TotalPoints'] = total_points['Points_x'] + total_points['Points_y']

#Get the team with the lowest total points in each season
lowest_points_per_season = total_points.groupby('Season_End_Year').apply(lambda x: x.loc[x['TotalPoints'].idxmin()]).reset_index(drop=True)

#Display the results
print("Teams with the Lowest Total Points Accumulated in Each Season (Including Last Season):")
print(lowest_points_per_season[['Season_End_Year', 'HomeTeam', 'Points_x', 'Points_y', 'TotalPoints']])


Teams with the Lowest Total Points Accumulated in Each Season (Including Last Season):
    Season_End_Year           HomeTeam  Points_x  Points_y  TotalPoints
0              1993  Nottingham Forest        22        18           40
1              1994       Swindon Town        19        11           30
2              1995       Ipswich Town        18         9           27
3              1996             Bolton        19        10           29
4              1997  Nottingham Forest        18        16           34
5              1998     Crystal Palace        11        22           33
6              1999  Nottingham Forest        16        14           30
7              2000            Watford        19         5           24
8              2001      Bradford City        19         7           26
9              2002     Leicester City        16        12           28
10             2003         Sunderland        11         8           19
11             2004       Leeds United        22 

As seen, Derby County holds the undesirable record for the lowest points ever accumulated in a Premier League season, accruing only 11 points. This achievement, or rather lack thereof, occurred during the 2007-2008 season.

Teams with the highest amount of points in each season

In [ ]:
#Calculate home and away points for each team in each season
home_points = df[df['FullTimeResult'].isin(['H', 'D'])].groupby(['Season_End_Year', 'HomeTeam'])['Points'].sum().reset_index()
away_points = df[df['FullTimeResult'].isin(['A', 'D'])].groupby(['Season_End_Year', 'AwayTeam'])['Points'].sum().reset_index()

#Merge home and away points, filling NaN with 0
total_points = pd.merge(home_points, away_points, how='outer', left_on=['Season_End_Year', 'HomeTeam'], right_on=['Season_End_Year', 'AwayTeam'])
total_points = total_points.fillna(0)

#Calculate total points for each team in each season
total_points['TotalPoints'] = total_points['Points_x'] + total_points['Points_y']

#Get the team with the highest total points in each season
highest_points_per_season = total_points.groupby('Season_End_Year').apply(lambda x: x.loc[x['TotalPoints'].idxmax()]).reset_index(drop=True)

#Display results
print("Teams with the Lowest Total Points Accumulated in Each Season (Including Last Season):")
print(highest_points_per_season[['Season_End_Year', 'HomeTeam', 'Points_x', 'Points_y', 'TotalPoints']])


Teams with the Lowest Total Points Accumulated in Each Season (Including Last Season):
    Season_End_Year         HomeTeam  Points_x  Points_y  TotalPoints
0              1993   Manchester Utd        47        37           84
1              1994   Manchester Utd        48        44           92
2              1995        Blackburn        53        36           89
3              1996   Manchester Utd        49        33           82
4              1997   Manchester Utd        41        34           75
5              1998          Arsenal        47        31           78
6              1999   Manchester Utd        46        33           79
7              2000   Manchester Utd        49        42           91
8              2001   Manchester Utd        47        33           80
9              2002          Arsenal        40        47           87
10             2003   Manchester Utd        50        33           83
11             2004          Arsenal        49        41           90
12 

Manchester City notably getting 100 points. The most ever in a season

In [ ]:
# Get the team with the highest total points in each season
title_winners = total_points.groupby('Season_End_Year')['HomeTeam'].apply(lambda x: x[total_points.loc[x.index, 'TotalPoints'].idxmax()]).reset_index()
title_winners.columns = ['Season_End_Year', 'Team']

# Count the number of titles for each team
title_counts = title_winners['Team'].value_counts().reset_index()
title_counts.columns = ['Team', 'count']

# Display the results
print("Teams with Title Counts:")
title_counts


Teams with Title Counts:


Team  count
0   Manchester Utd     13
1  Manchester City      7
2          Chelsea      5
3          Arsenal      3
4        Blackburn      1
5   Leicester City      1
6        Liverpool      1

In [ ]:
# Create a pie chart
fig = px.pie(title_counts, names='Team', values='count', title='Premier League Title Counts by Team')
fig.show()

Goals

Analyze the distribution of goals scored in each season

In [ ]:
#Convert 'Date' to datetime data type
df['Date'] = pd.to_datetime(df['Date'])

#Calculate total goals scored in each season
total_goals_by_season_home = df.groupby('Season_End_Year')['HomeGoals'].sum().reset_index()
total_goals_by_season_away = df.groupby('Season_End_Year')['AwayGoals'].sum().reset_index()

#Calculate total goals (home + away)
total_goals_by_season = total_goals_by_season_home.merge(total_goals_by_season_away, on='Season_End_Year', suffixes=('_home', '_away'))
total_goals_by_season['TotalGoals'] = total_goals_by_season['HomeGoals'] + total_goals_by_season['AwayGoals']

#Create plot
fig = px.bar(total_goals_by_season, x='Season_End_Year', y='TotalGoals', text='TotalGoals',
             labels={'TotalGoals': 'Total Goals'}, title='Total Goals Scored in Each Season')

#Set labels and title
fig.update_layout(xaxis_title='Season End Year', yaxis_title='Total Goals', uniformtext_minsize=8, uniformtext_mode='hide')

#Show plot
fig.show()


In [ ]:
#Calculate the total goals across all seasons
total_goals_all_seasons = total_goals_by_season['TotalGoals'].sum()

#Display the total amount of goals
print("Total Goals Scored Across All Seasons:", total_goals_all_seasons)


Total Goals Scored Across All Seasons: 32072


Relegations

In [ ]:
# Calculate home and away points for each team in each season
home_points = df[df['FullTimeResult'].isin(['H', 'D'])].groupby(['Season_End_Year', 'HomeTeam'])['Points'].sum().reset_index()
away_points = df[df['FullTimeResult'].isin(['A', 'D'])].groupby(['Season_End_Year', 'AwayTeam'])['Points'].sum().reset_index()

# Merge home and away points, filling NaN with 0
total_points = pd.merge(home_points, away_points, how='outer', left_on=['Season_End_Year', 'HomeTeam'], right_on=['Season_End_Year', 'AwayTeam'])
total_points = total_points.fillna(0)

# Calculate total points for each team in each season
total_points['TotalPoints'] = total_points['Points_x'] + total_points['Points_y']

# Get the 3 teams with the lowest total points in each season
lowest_points_per_season = (
    total_points.sort_values(['Season_End_Year', 'TotalPoints'])
    .groupby('Season_End_Year').head(3)
    .reset_index(drop=True)
)

In [ ]:
#Count the occurrences of each team in the list of teams with the lowest points
teams_count = lowest_points_per_season['HomeTeam'].value_counts().reset_index()
teams_count.columns = ['Team', 'Count']

#Get the top 5 teams that appear the most
top_5_teams = teams_count.head(5)

print("Top 5 Teams With the most Relegations:")
top_5_teams

Top 5 Teams With the most Relegations:


Team  Count
0  Norwich City      6
1     West Brom      5
2       Watford      4
3        Fulham      4
4    Sunderland      4